examples/step-57/step-57.cc

```diff
@@ -0,0 +1,878 @@
+      return 1.0;
+
+    return 0;
+  }
+
+  template <int dim>
+  void BoundaryValues<dim>::vector_value (const Point<dim> &p,
+                                          Vector<double>   &values) const
+  {
+    for (unsigned int c = 0; c < this->n_components; ++c)
+      values(c) = BoundaryValues<dim>::value (p, c);
+  }
+
+
+  // @sect3{BlockSchurPreconditioner for Navier Stokes equations}
+  //
+  // The block
+  // Schur complement preconditioner is defined in this part. As discussed in
+  // the introduction, the preconditioner in Krylov iterative methods is
+  // implemented as a matrix-vector product operator. In practice, the Schur
+  // complement preconditioner is decomposed as a product of three matrices(as
+  // presented in the first section).  The $\tilde{A}^{-1}$ in the first factor
+  // involves a solve for the linear system $\tilde{A}x=b$. Here we solve
+  // this system via a direct solver for simplicity. The computation involved
+  // in the second factor is a simple matrix-vector multiplication. The Schur
+  // complement $\tilde{S}$ can be well approximated by the pressure mass
+  // matrix and its inverse can be obtained through an inexact solver. Because
+  // the pressure mass matrix is symmetric and positive definite, we can use
+  // CG to solve the corresponding linear system.
+  //
+  template <class PreconditionerMp>
+  class BlockSchurPreconditioner : public Subscriptor
+  {
+  public:
+    BlockSchurPreconditioner (double                                     gamma,
+                              double                                     viscosity,
+                              const BlockSparseMatrix<double>            &S,
+                              const SparseMatrix<double>                 &P,
+                              const PreconditionerMp                     &Mppreconditioner
+                             );
+
+    void vmult (BlockVector<double>       &dst,
+                const BlockVector<double> &src) const;
+
+  private:
+    const double gamma;
+    const double viscosity;
+    const BlockSparseMatrix<double> &stokes_matrix;
+    const SparseMatrix<double> &pressure_mass_matrix;
+    const PreconditionerMp &mp_preconditioner;
+    SparseDirectUMFPACK A_direct;
+  };
+
+  template <class PreconditionerMp>
+  BlockSchurPreconditioner<PreconditionerMp>::
+  BlockSchurPreconditioner (double                           gamma,
+                            double                           viscosity,
+                            const BlockSparseMatrix<double>  &S,
+                            const SparseMatrix<double>       &P,
+                            const PreconditionerMp           &Mppreconditioner)
+    :
+    gamma                (gamma),
+    viscosity            (viscosity),
+    stokes_matrix        (S),
+    pressure_mass_matrix (P),
+    mp_preconditioner    (Mppreconditioner)
+  {
+    A_direct.initialize(stokes_matrix.block(0,0));
```

Comment:
I think it would be worthwhile saying in the introduction that the initialize call here computes the inverse matrix, and that doing this here makes sure that every application of the preconditioner then no longer requires the computation of the matrix factors.